keydata.go

```diff
@@ -482,110 +504,180 @@ func (d *KeyData) snapModelAuthKey(auxKey AuxiliaryKey) ([]byte, error) {
 	return hmacKey, nil
 }
 
-func (d *KeyData) updatePassphrase(payload, oldKey []byte, passphrase string, kdfOptions *KDFOptions, kdf KDF) error {
-	handler := handlers[d.data.PlatformName]
-	if handler == nil {
-		return ErrNoPlatformHandlerRegistered
+func (d *KeyData) derivePassphraseKeys(passphrase string, kdf KDF) (key, iv, auth []byte, err error) {
+	if d.data.PassphraseParams == nil {
+		return nil, nil, nil, errors.New("no passphrase params")
 	}
 
-	if kdfOptions == nil {
-		var defaultOptions KDFOptions
-		kdfOptions = &defaultOptions
+	params := d.data.PassphraseParams
+	if params.KDF.Type != kdfType {
+		// Only Argon2i is supported
+		return nil, nil, nil, fmt.Errorf("unexpected intermediate KDF type \"%s\"", params.KDF.Type)
+	}
+	if params.DerivedKeySize < 0 {
+		return nil, nil, nil, fmt.Errorf("invalid derived key size (%d bytes)", params.DerivedKeySize)
+	}
+	if params.EncryptionKeySize < 0 || params.EncryptionKeySize > 32 {
+		// The key size can't be larger than 32 with the supported cipher
+		return nil, nil, nil, fmt.Errorf("invalid encryption key size (%d bytes)", params.EncryptionKeySize)
+	}
+	if params.AuthKeySize < 0 {
+		return nil, nil, nil, fmt.Errorf("invalid auth key size (%d bytes)", params.AuthKeySize)
```

Comment:
should/can some of these checks be tested?

Reply:
I added some tests that use `RecoverKeysWithPassphrase`.